### DISCUSSION TITLE
Confused by user info, timestamps, and urls in Commit objects

### DISCUSSION THREAD

**Discussion by penyuan** [community/community#24391] [Apps, API and Webhooks]
<p>Dear GitHub Support Community,</p>
<p>As I continue to learn the GitHub GraphQL API, I have used the following sample query with the <a href="https://github.com/graphql-python/gql" rel="nofollow noopener">gql library for Python</a>:</p>
<pre><code class="lang-auto">query {
  repository(owner: "github", name: "linguist") {
    refs(query: "dev", refPrefix: "refs/heads/", first: 1) {
      edges {
        node {
          target {
            ... on Commit {
              history(first: 5, after: null) {
                edges {
                  node {
                    oid
                    commitUrl
                    url
                    authoredByCommitter
                    authoredDate
                    author {
                      user {
                        email
                        login
                        name
                        twitterUsername
                      }
                      date
                    }
                    committedDate
                    committer {
                      user {
                        email
                        login
                        name
                        twitterUsername
                      }
                      date
                    }
                  }
                }
                pageInfo {
                  hasNextPage
                  endCursor
                }
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p>And here is the API response:</p>
<details>
<summary>
Click to expand API response</summary>
<pre><code class="lang-auto">{
  "data": {
    "repository": {
      "refs": {
        "edges": [
          {
            "node": {
              "target": {
                "history": {
                  "edges": [
                    {
                      "node": {
                        "oid": "994bc1f13560661f94c880845c967c91116e2f87",
                        "commitUrl": "https://github.com/github/linguist/commit/994bc1f13560661f94c880845c967c91116e2f87",
                        "url": "https://github.com/github/linguist/commit/994bc1f13560661f94c880845c967c91116e2f87",
                        "authoredByCommitter": false,
                        "authoredDate": "2017-05-03T13:49:26Z",
                        "author": {
                          "user": {
                            "email": "",
                            "login": "lildude",
                            "name": "Colin Seymour",
                            "twitterUsername": null
                          },
                          "date": "2017-05-03T14:49:26+01:00"
                        },
                        "committedDate": "2017-05-03T13:49:26Z",
                        "committer": {
                          "user": null,
                          "date": "2017-05-03T14:49:26+01:00"
                        }
                      }
                    },
                    {
                      "node": {
                        "oid": "44f03e64c1f79dc03f67a1e63192c59932ed7147",
                        "commitUrl": "https://github.com/github/linguist/commit/44f03e64c1f79dc03f67a1e63192c59932ed7147",
                        "url": "https://github.com/github/linguist/commit/44f03e64c1f79dc03f67a1e63192c59932ed7147",
                        "authoredByCommitter": false,
                        "authoredDate": "2017-04-29T09:15:39Z",
                        "author": {
                          "user": {
                            "email": "gardnerjohng@gmail.com",
                            "login": "Alhadis",
                            "name": "John Gardner",
                            "twitterUsername": null
                          },
                          "date": "2017-04-29T19:15:39+10:00"
                        },
                        "committedDate": "2017-04-29T09:15:39Z",
                        "committer": {
                          "user": {
                            "email": "paul.chaignon@gmail.com",
                            "login": "pchaigno",
                            "name": "Paul Chaignon",
                            "twitterUsername": "pchaigno"
                          },
                          "date": "2017-04-29T11:15:39+02:00"
                        }
                      }
                    },
                    {
                      "node": {
                        "oid": "4166f2e89d3e0d0d21e67ac5677a56a4efaa8221",
                        "commitUrl": "https://github.com/github/linguist/commit/4166f2e89d3e0d0d21e67ac5677a56a4efaa8221",
                        "url": "https://github.com/github/linguist/commit/4166f2e89d3e0d0d21e67ac5677a56a4efaa8221",
                        "authoredByCommitter": false,
                        "authoredDate": "2017-04-28T23:20:22Z",
                        "author": {
                          "user": {
                            "email": "",
                            "login": "jelder",
                            "name": "Jacob Elder",
                            "twitterUsername": null
                          },
                          "date": "2017-04-28T19:20:22-04:00"
                        },
                        "committedDate": "2017-04-28T23:20:22Z",
                        "committer": {
                          "user": {
                            "email": "",
                            "login": "lildude",
                            "name": "Colin Seymour",
                            "twitterUsername": null
                          },
                          "date": "2017-04-28T16:20:22-07:00"
                        }
                      }
                    },
                    {
                      "node": {
                        "oid": "1a8f19c6f28b4ad9acb20351a323aa1fc8e4b26e",
                        "commitUrl": "https://github.com/github/linguist/commit/1a8f19c6f28b4ad9acb20351a323aa1fc8e4b26e",
                        "url": "https://github.com/github/linguist/commit/1a8f19c6f28b4ad9acb20351a323aa1fc8e4b26e",
                        "authoredByCommitter": false,
                        "authoredDate": "2017-04-28T21:02:38Z",
                        "author": {
                          "user": {
                            "email": "gardnerjohng@gmail.com",
                            "login": "Alhadis",
                            "name": "John Gardner",
                            "twitterUsername": null
                          },
                          "date": "2017-04-29T07:02:38+10:00"
                        },
                        "committedDate": "2017-04-28T21:02:38Z",
                        "committer": {
                          "user": {
                            "email": "",
                            "login": "lildude",
                            "name": "Colin Seymour",
                            "twitterUsername": null
                          },
                          "date": "2017-04-28T14:02:38-07:00"
                        }
                      }
                    },
                    {
                      "node": {
                        "oid": "c0e242358af6a3ad0ace66e29b480d3f507917cb",
                        "commitUrl": "https://github.com/github/linguist/commit/c0e242358af6a3ad0ace66e29b480d3f507917cb",
                        "url": "https://github.com/github/linguist/commit/c0e242358af6a3ad0ace66e29b480d3f507917cb",
                        "authoredByCommitter": false,
                        "authoredDate": "2017-04-26T22:31:36Z",
                        "author": {
                          "user": {
                            "email": "santi@mola.io",
                            "login": "smola",
                            "name": "Santiago M. Mola",
                            "twitterUsername": null
                          },
                          "date": "2017-04-27T00:31:36+02:00"
                        },
                        "committedDate": "2017-04-26T22:31:36Z",
                        "committer": {
                          "user": {
                            "email": "",
                            "login": "lildude",
                            "name": "Colin Seymour",
                            "twitterUsername": null
                          },
                          "date": "2017-04-26T15:31:36-07:00"
                        }
                      }
                    }
                  ],
                  "pageInfo": {
                    "hasNextPage": true,
                    "endCursor": "994bc1f13560661f94c880845c967c91116e2f87 4"
                  }
                }
              }
            }
          }
        ]
      }
    }
  }
}
</code></pre>
</details>
<p>I am confused about the seemingly subtle differences between some of the fields, specifically:</p>
<ol>
<li>What is the difference between <code>author</code> and <code>committer</code> in the API response? Do their meanings differ in Git vs GitHub? For example, <a href="https://github.com/github/linguist/commit/44f03e64c1f79dc03f67a1e63192c59932ed7147" rel="nofollow noopener">the second commit’s page</a> shows that " <a href="https://github.com/github/linguist/commits?author=Alhadis" rel="nofollow noopener">Alhadis</a> authored and <a href="https://github.com/github/linguist/commits?author=pchaigno" rel="nofollow noopener">pchaigno</a> committed on Apr 29, 2017". What real-life scenarios would lead to this? As a beginner Git user I’ve never encountered anything like this.</li>
<li>From what I can tell, a <code>Commit</code> from the API response can include multiple timestamps such as <code>authoredDate</code>, <code>committedDate</code>, and the <code>date</code> under <code>author</code> and <code>committer</code>. They are identical in the response I got from my sample query above, but what would lead to them being different? And which one is shown on a commit’s <code>commitUrl</code> page?</li>
<li>On that note, what is the difference between <code>commitUrl</code> and <code>url</code> in a <code>Commit</code>?</li>
<li>In the first commit of the API response I got (commit <a href="https://github.com/github/linguist/commit/994bc1f13560661f94c880845c967c91116e2f87" rel="nofollow noopener">994bc1f</a>), it says <code>authoredByCommitter</code> is <code>False</code> <em>and</em> that the <code>committer</code> is <code>null</code>! (there is only information on the <code>author</code>) Why is this? <em>How can there be no committer</em>? If I follow the <code>commitUrl</code> I see that “<a href="https://github.com/github/linguist/commits?author=lildude" rel="nofollow noopener">lildude</a> committed on May 3, 2017” so presumably user <a href="https://github.com/github/linguist/commits?author=lildude" rel="nofollow noopener">lildude</a> should be the <code>committer</code> and <code>authoredByCommitter</code> should be <code>True</code>?</li>
</ol>
<p>Sorry for all the questions, hope someone can help elucidate these details! Thank you! :grin:</p>

**Comment by francisfuzz** [comment, marked as answer]
<aside class="quote no-group" data-username="penyuan" data-post="1" data-topic="125159">
<div class="title">
<div class="quote-controls"></div>
 penyuan:</div>
<blockquote>
<p>What is the difference between <code>author</code> and <code>committer</code> in the API response? Do their meanings differ in Git vs GitHub? For example, <a href="https://github.com/github/linguist/commit/44f03e64c1f79dc03f67a1e63192c59932ed7147">the second commit’s page</a> shows that " <a href="https://github.com/github/linguist/commits?author=Alhadis">Alhadis </a> authored and <a href="https://github.com/github/linguist/commits?author=pchaigno">pchaigno</a> committed on Apr 29, 2017". What real-life scenarios would lead to this? As a beginner Git user I’ve never encountered anything like this.</p>
</blockquote>
</aside>
<p>GitHub uses the metadata created by Git in a commit to present the respective author and committer information for a commit.</p>
<p>When using Git on the command line, it’s possible to specify an author of the commit who is not the committer themselves. This practice was common before pull requests and Git-hosting providers existed, where a contributor would submit a patch file to the mailing list and a committer would review the patch before committing it to the repository on behalf of the author.</p>
<p>Drawing from the <a href="https://git-scm.com/book/en/v2/Git-Basics-Viewing-the-Commit-History">Git SCM - <code>2.3 Git Basics - Viewing the Commit History</code></a>:</p>
<blockquote>
<p>You may be wondering what the difference is between  <em>author</em>  and  <em>committer</em> . The author is the person who originally wrote the work, whereas the committer is the person who last applied the work. So, if you send in a patch to a project and one of the core members applies the patch, both of you get credit — you as the author, and the core member as the committer.</p>
</blockquote>
<aside class="quote no-group" data-username="penyuan" data-post="1" data-topic="125159">
<div class="title">
<div class="quote-controls"></div>
 penyuan:</div>
<blockquote>
<ul>
<li>From what I can tell, a <code>Commit</code> from the API response can include multiple timestamps such as <code>authoredDate</code> , <code>committedDate</code> , and the <code>date</code> under <code>author</code> and <code>committer</code> . They are identical in the response I got from my sample query above, but what would lead to them being different? And which one is shown on a commit’s <code>commitUrl</code> page?</li>
</ul>
</blockquote>
</aside>
<p>Extending from the earlier explanation, it’s possible for an author to write the changes at a different time than when they’re committed. <code>committer.date</code> is used (though that’s subject to change at anytime––we suggest using the API to check based on what you’re looking for).</p>
<aside class="quote no-group" data-username="penyuan" data-post="1" data-topic="125159">
<div class="title">
<div class="quote-controls"></div>
 penyuan:</div>
<blockquote>
<ul>
<li>On that note, what is the difference between <code>commitUrl</code> and <code>url</code> in a <code>Commit</code> ?</li>
</ul>
</blockquote>
</aside>
<p>Looking at the <a href="https://docs.github.com/en/graphql/reference/objects#commit"><code>Commit</code></a> reference documentation:</p>
<ul>
<li>
<code>commitUrl</code>  ( <code>URI!</code> ) The HTTP URL for this Git object.</li>
<li>
<code>url</code>  ( <code>URI!</code> ) The HTTP URL for this commit.</li>
</ul>
<p>The <a href="https://docs.github.com/en/graphql/reference/interfaces#gitobject"><code>GitObject</code></a> interface is implemented by <code>Commit</code>. Thus, <code>commitUrl</code> is shared among its other implementations (like <code>Blob</code>, <code>Tag</code>, and <code>Tree</code>) while having its own <code>url</code>. In any case, you’re welcome to use either interchangeably.</p>
<aside class="quote no-group" data-username="penyuan" data-post="1" data-topic="125159">
<div class="title">
<div class="quote-controls"></div>
 penyuan:</div>
<blockquote>
<ul>
<li>In the first commit of the API response I got (commit <a href="https://github.com/github/linguist/commit/994bc1f13560661f94c880845c967c91116e2f87">994bc1f</a>), it says <code>authoredByCommitter</code> is <code>False</code> <em>and</em> that the <code>committer</code> is <code>null</code> ! (there is only information on the <code>author</code> ) Why is this? <em>How can there be no committer</em> ? If I follow the <code>commitUrl</code> I see that “<a href="https://github.com/github/linguist/commits?author=lildude">lildude</a> committed on May 3, 2017” so presumably user <a href="https://github.com/github/linguist/commits?author=lildude">lildude</a> should be the <code>committer</code> and <code>authoredByCommitter</code> should be <code>True</code> ?</li>
</ul>
</blockquote>
</aside>
<p><code>authoredByCommitter</code> is <code>False</code> because the author and committer information don’t match.</p>
<p>We can inspect that particular commit by making a request to either of these endpoints:</p>
<ul>
<li><a href="https://docs.github.com/en/rest/reference/git#get-a-commit">Git Database API: Get a commit</a></li>
<li><a href="https://docs.github.com/en/rest/reference/repos#get-a-commit">Repositories API: Get a commit</a></li>
</ul>
<p>I made the following calls and extracted the respective <code>author</code> and <code>committer</code> fields here:</p>
<ul>
<li><a href="https://api.github.com/repos/github/linguist/commits/994bc1f13560661f94c880845c967c91116e2f87">https://api.github.com/repos/github/linguist/commits/994bc1f13560661f94c880845c967c91116e2f87</a></li>
</ul>
<pre><code class="lang-json">// truncated for brevity
{
  // ...
  "author": {
    "name": "Colin Seymour",
    "email": "colin@github.com",
    "date": "2017-05-03T13:49:26Z"
  },
  "committer": {
    "name": "GitHub",
    "email": "noreply@github.com",
    "date": "2017-05-03T13:49:26Z"
  },
  // ...
}
</code></pre>
<ul>
<li><a href="https://api.github.com/repos/github/linguist/git/commits/994bc1f13560661f94c880845c967c91116e2f87">https://api.github.com/repos/github/linguist/git/commits/994bc1f13560661f94c880845c967c91116e2f87</a></li>
</ul>
<pre><code class="lang-auto">// truncated for brevity
{
  // ...
  "commit": {
    "author": {
      "name": "Colin Seymour",
      "email": "colin@github.com",
      "date": "2017-05-03T13:49:26Z"
    },
    "committer": {
      "name": "GitHub",
      "email": "noreply@github.com",
      "date": "2017-05-03T13:49:26Z"
    },
  // ... 
}
</code></pre>
<p>In cases like these, it’s possible that this commit was created through the GitHub web flow (like editing a file and committing the changes, or pressing the merge button to merge the changes proposed in a pull request). Building from that context, <a href="https://docs.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user">this guide</a> explains why commits may not be linked to any user.</p>
<p>Does that help explain things?</p>

**Comment by penyuan** [comment]
<aside class="quote group-github-employees quote-modified" data-username="francisfuzz" data-post="2" data-topic="125159">
<div class="title">
<div class="quote-controls"></div>
 francisfuzz:</div>
<blockquote>
<p>When using Git on the command line, it’s possible to specify an author of the commit who is not the committer themselves. This practice was common before pull requests and Git-hosting providers existed…</p>
</blockquote>
</aside>
<p>Ah I didn’t know that. Great to know that there’s flexibility in how to provide attribution and credit in Git.</p>
<blockquote>
<p><code>committer.date</code> is used (though that’s subject to change at anytime––we suggest using the API to check based on what you’re looking for).</p>
</blockquote>
<blockquote>
<p>The <a href="https://docs.github.com/en/graphql/reference/interfaces#gitobject" rel="nofollow noopener"> <code>GitObject</code> </a> interface is implemented by <code>Commit</code> . Thus, <code>commitUrl</code> is shared among its other implementations (like <code>Blob</code> , <code>Tag</code> , and <code>Tree</code> ) while having its own <code>url</code> . In any case, you’re welcome to use either interchangeably.</p>
</blockquote>
<p>OK.</p>
<blockquote>
<p>In cases like these, it’s possible that this commit was created through the GitHub web flow (like editing a file and committing the changes, or pressing the merge button to merge the changes proposed in a pull request).</p>
</blockquote>
<p>It’s a bit counterintuitive to me, since I’d think in this case the user would be attributed to as the <code>committer</code> instead of <code>author</code>, but that’s OK. As long as I know which one’s which that’s great.</p>
<blockquote>
<p>Does that help explain things?</p>
</blockquote>
<p>Yes, thank you! :pray:t5:</p>
